### PR TITLE
nerdctl wait command

### PR DESCRIPTION
**Pull request by fahedouch**
Signed-off-by: fahed dorgaa <fahed.dorgaa@gmail.com>

### CODE REVIEW

File: container_wait.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	client, ctx, cancel, err := newClient(clicontext)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			if err := waitContainer(clicontext, ctx, client, found.Container.ID()); err != nil {
```
Comment (AkihiroSuda) [inline container_wait.go:52]: `nerdctl wait foo bar` may not be able to print the status of `bar` when `bar` exits and gets removed before `foo` exits.
So `task.Wait()` should be called in goroutines.

File: container_wait.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (AkihiroSuda) [inline container_wait.go:1]: The entire file can be just merged into `wait.go`

File: wait_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+func TestWaitExit0(t *testing.T) {
+	const (
+		testContainerName = "nerdctl-test-wait"
+	)
+	base := testutil.NewBase(t)
+	defer base.Cmd("rm", "-f", testContainerName).Run()
+
+	base.Cmd("run", "-d", "--name", testContainerName, testutil.AlpineImage, "sleep", "10").AssertOK()
+
+	base.Cmd("wait", testContainerName).AssertOK()
```
Comment (AkihiroSuda) [inline wait_test.go:34]: `wait` should be tested with multiple containers

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by AkihiroSuda** [comment]
Could you also update README (https://github.com/containerd/nerdctl#unimplemented-docker-commands) ?

**Comment by fahedouch** [comment]
> Could you also update README (https://github.com/containerd/nerdctl#unimplemented-docker-commands) ?

@AkihiroSuda  if all is ok then I will squash commit!

File: container_wait.go (hunk 1 above)
Comment (fahedouch) [reply container_wait.go]: Done

File: README.md [hunk 4]
```diff
@@ -477,6 +477,11 @@ Usage: `nerdctl start [OPTIONS] CONTAINER [CONTAINER...]`
 
 Unimplemented `docker start` flags: `--attach`, `--checkpoint`, `--checkpoint-dir`, `--detach-keys`, `--interactive`
 
+### :whale: nerdctl wait
```
Comment (AkihiroSuda) [inline README.md:480]: Could you run `doctor README.md` to sync the TOC

https://github.com/thlorenz/doctoc

File: container.go [hunk 5]
```diff
@@ -36,6 +36,7 @@ var containerCommand = &cli.Command{
 		startCommand,
 		killCommand,
 		pauseCommand,
+		containerWaitCommand,
```
Comment (AkihiroSuda) [inline container.go:39]: waitCommand can be used directly

File: wait.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+var waitCommand = &cli.Command{
+	Name:         "wait",
+	Usage:        "Block until one or more containers stop, then print their exit codes.",
+	Action:       containerWaitAction,
+	BashComplete: waitBashComplete,
+}
+
+var containerWaitCommand = &cli.Command{
```
Comment (AkihiroSuda) [inline wait.go:38]: Unneeded

File: wait.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+			return nil
+		},
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(clicontext.Args().Slice()))
+	var errs []error
+	for _, req := range clicontext.Args().Slice() {
+		go func(req string) {
+			defer wg.Done()
+			n, err := walker.Walk(ctx, req)
```
Comment (AkihiroSuda) [inline wait.go:72]: Walk isn't thread-safe, only task.Wait and Fprintf should happen in goroutines

File: wait.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	walker := &containerwalker.ContainerWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			if err := waitContainer(clicontext, ctx, client, found.Container.ID()); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+	var wg sync.WaitGroup
```
Comment (AkihiroSuda) [inline wait.go:66]: errgroup

File: wait_test.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+func TestWaitExit0(t *testing.T) {
+	const (
+		testContainerName1 = "nerdctl-test-wait-1"
+		testContainerName2 = "nerdctl-test-wait-2"
+	)
+	base := testutil.NewBase(t)
+	defer base.Cmd("rm", "-f", testContainerName1, testContainerName2).Run()
+
+	base.Cmd("run", "-d", "--name", testContainerName1, testutil.AlpineImage, "sleep", "10").AssertOK()
```
Comment (AkihiroSuda) [inline wait_test.go:33]: Sleep 10 is too long, 3 secs would be enough

File: wait_test.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+	base.Cmd("run", "-d", "--name", testContainerName2, testutil.AlpineImage, "sleep", "5").AssertOK()
```
Comment (AkihiroSuda) [inline wait_test.go:35]: At least one of the command should exit with non-zero

File: wait_test.go [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+	base.Cmd("wait", testContainerName1, testContainerName2).AssertOK()
```
Comment (AkihiroSuda) [inline wait_test.go:37]: AssertOut

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: container.go (hunk 5 above)
Comment (fahedouch) [reply container.go:39]: I know but we have too docker container wait. Should i ignore it ?

Comment (AkihiroSuda) [reply container.go:39]: Please see other commands such as pauseCommand, and follow the coding style

File: wait.go (hunk 8 above)
Comment (fahedouch) [reply wait.go:66]: I think that the errgroup is not made to get all errors but only the first occured error which is not the case on docker wait. So i think it is better to keep using `Waitgroup`

File: container.go (hunk 5 above)
Comment (fahedouch) [reply container.go:39]: Yep I see

File: wait.go (hunk 8 above)
Comment (AkihiroSuda) [reply wait.go:66]: Either is fine to me

File: wait.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return err
+	}
+
+	task, err := container.Task(ctx, cio.Load)
+	if err != nil {
+		return err
+	}
+
+    var statusC <-chan containerd.ExitStatus
```
Comment (AkihiroSuda) [inline wait.go:83]: unneeded line

File: wait.go [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    g.Go(func() error {
+        statusC, err = task.Wait(ctx)
```
Comment (AkihiroSuda) [inline wait.go:86]: err has to be defined inside `g.Go(func()error{ ... }`

File: wait.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+			}
+	}
+
+    if err := g.Wait(); err != nil {
+            return err
+    }
+	return nil
+}
+
+func waitContainer(clicontext *cli.Context, ctx context.Context, client *containerd.Client, id string, g *errgroup.Group) error {
```
Comment (AkihiroSuda) [inline wait.go:72]: nit: the standard convention is to have `context.Context` before `*cli.Context`

File: wait_test.go [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+	base.Cmd("run", "-d", "--name", testContainerName1, testutil.AlpineImage, "sleep", "3").AssertOK()
+
+	base.Cmd("run", "-d", "--name", testContainerName2, testutil.AlpineImage, "sleep", "4").AssertOK()
+
+	base.Cmd("wait", testContainerName1, testContainerName2).AssertOut(expected)
+
+}
+
+func TestWaitExit137(t *testing.T) {
```
Comment (AkihiroSuda) [inline wait_test.go:45]: The number in the test name doesn't seem consistent with the actual code.

Comment (AkihiroSuda) [reply wait_test.go:45]: You can just squash `TestWaitExit0` and `TestWaitExit137` into a single `TestWait`, and check `0\n0\n123\n`

File: wait.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+   limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+
+	//"sync"
```
Comment (AkihiroSuda) [inline wait.go:23]: delete

File: wait_test.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+		testContainerName2 = "nerdctl-test-wait-2"
+		testContainerName3 = "nerdctl-test-wait-3"
+	)
+
+	const expected = `0
+0
+123
+`
+	base := testutil.NewBase(t)
+	defer base.Cmd("rm", "-f", testContainerName1, testContainerName2).Run()
```
Comment (AkihiroSuda) [inline wait_test.go:37]: Also rm testContainerName3

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**